### DISCUSSION TITLE
Loot drop tracking via AHK

### DISCUSSION THREAD

**Discussion by GrimDawn-max** [Lailloken/Exile-UI#105] [Ideas & Feedback]
Would it be possible to track the loot drops you get that are stored in player inventory (not hideout stash). Only those items that you actually pick up, not just hover over (that would be crazy). The script would write the contents of inventory to a log file that a user could then use for their own purposes. Personally I'd use it to keep track of how many of each thing I get. Others might use for other things. Right now that's all a manual process.

**Comment by Lailloken** [comment]
Yeah, I think that would be doable. However, there are apps that specialize in tracking loot and currency which are far more advanced than what I could implement at the moment. Have you tried those, e.g. exilence-next?

What I could do is implement a feature that logs every item that is ctrl-clicked from the inventory to the stash. I wanted to implement advanced stash features using the stash/trade APIs for some time now but haven't gotten to that yet. I don't know if it's possible to implement these things with AHK or if the inventory is even accessible this way.

**Reply by GrimDawn-max** [reply]
I haven't tested exilence-next. It looks like it collects similar data but is it looking at your HO stash or just the inventory stash ? I want the inventory stash at the end of each map. I don't particularly care about the cash value. I'll give it a try and see if it satisfy the immediate need but tbh I'd prefer to use an all in one tool like yours that does multiple things rather than having to use too many third party apps. It's already out of control. That's why yours is so handy. It's the swiss army knife of utilities for POE IMHO. 

Your suggestion about tracking the ctrl+clicked items between inventory and stash sounds exactly what I'm after. Let me know if you get the time to implement something like that and I'll test it out.

**Reply by Lailloken** [reply]
> It's the swiss army knife of utilities for POE IMHO

Nice to hear that, much appreciated.

> Let me know if you get the time to implement something like that and I'll test it out.

I had to switch PCs recently and lost some progress on a new feature I was testing. I'll let you know when I start working on your suggestion.

**Reply by GrimDawn-max** [reply]
One more thought - when in Labs I often forget how many of the dark shrines I've done/missed. POE Lab has a count of the number of dark shrines for each lab daily. Would there be a way to capture that and then check the Client text file for the number of dark shrine comments in a run. That would tell me if I've got them all or I'm missing 1 or more.

**Reply by Lailloken** [reply]
Yeah, I had plans to revisit lab features after implementing client-txt tracking for the guide overlays. I have not looked into it yet, but it's definitely something I want to improve/expand.

**Reply by Lailloken** [reply]
So, I just started planning the drop tracker to see what it would roughly look like:
- by parsing the client.txt, I can kind of reliably track which maps are being done: each opened map has a seed, so every time the script parses a new seed it will be able to associate ctrl-clicked items from that point onward with that specific map

- having a seed means it can still track items reliably even if the player only plays one kind of map and target-farms it
- small problem is that it will track every item in between maps, e.g. if you trade with someone and stash the currency you get from them; so you would have to remember to drag items you don't want to track, instead of ctrl-clicking
- since planning is still in an early phase, I don't know how extensive this feature will be: it may only dump the data into a log, or I may implement a full UI to access the data and display it nicely (depends on how interesting I find this feature, and how much I think I can add to it)
- if I do end up creating a UI, I could also easily add a pause button
- there were some feature suggestions in the past I could include in the tracked data (e.g. map-time, map-time vs hideout-time, etc.)

**Reply by GrimDawn-max** [reply]
I'd be ok with just dumping the inventory items in a file. Players, such as myself, can then use that however they want. It's be the simplest thing for you to do. Players can then edit the dump file and remove items they don't want to include, or just write a script to exclude them in any further data analysis. 

If you did go the UI route maybe just include a remove button so it excludes any item in the inventory stash a user doesn't want to include in any calculation (I'd still create a dump file for those who still want to play with the info though).

Map timer would be cool but afaik there is no way to definitively know when you've completed a map, at least not via the client.txt file. It doesn't write anything which is odd and annoying. Most times when in a map you get a message about a mission or a craft received by killing the boss but sometimes you get nothing too so that isn't a totally reliable method to capture map completion either. If you could capture when a player kills the map boss that might work except some players rush to kill the boss and then complete the rest of the map. That one is going to be a pain unless you create your own map "end" definition and have a disclaimer so everyone is aware of how the calculation is done.

**Reply by Lailloken** [reply]
I think I'll track the net time one spends in a map with a specific seed. If one leaves the map, the timer will pause. And entering a different map (or same map with a different seed) will trigger the previous one as completed, save the loot, and start tracking for the next map. At least that's the theory.

**Reply by Lailloken** [reply]
I have a rough, initial test-version ready that includes a general mapping tracker, optional loot tracker, interface to view the logged data and export it as a csv-file. Do you want to beta-test it? I didn't and will not have much time this week, so the proper version will probably not be released before next week.

**Reply by GrimDawn-max** [reply]
> I have a rough, initial test-version ready that includes a general mapping tracker, optional loot tracker, interface to view the logged data and export it as a csv-file. Do you want to beta-test it? I didn't and will not have much time this week, so the proper version will probably not be released before next week.

I just finished building a test box so perfect timing. I can slap this on and see how things go, probably tomorrow or Friday. I'll let you know if I run across any issues or need help figuring it out. If something comes up and I don't get to test it I'm happy to wait for the final version when it comes out as well. Can you drop a zip file somewhere here so I can pick it up ?

**Reply by Lailloken** [reply]
https://github.com/Lailloken/Lailloken-UI/archive/refs/heads/dev.zip

This should be run either from clean install, or you can copy it into a separate, backed up LLK-UI folder. The settings menu has a new section 'mapping tracker', where the feature has to be enabled -- I implemented the loot tracker as an optional sub-feature because I might want to expand the mapping tracker in the future.

Enabling the mapping tracker will display a new button (similar to notepad, delve, etc.) that can be freely placed, and a small panel in the corner between exp-bar and open inventory. This panel displays the current map being tracked and the map-time. It should auto-detect new maps by parsing the client.txt for a new map-seed. Leaving the map will pause the timer, re-entering will resume it. I haven't implemented side-area detection (lab trials, abysses, vaal, etc.), so they should also trigger the 'outside of map' detection and pause the timer. The panel can be repositioned in the settings menu (panel offset).

If the loot tracker is enabled, ctrl-clicking items will log them and list them in this panel, as long as the tracker is active (as a preview and to double-check; stash image-check has to be setup). Items on this list can be clicked to remove them. Pressing ESC while the list is visible will change the panel back to the small form, only showing map name and timer. You can long-click the map name to manually trigger a map as completed -- this is only needed if you know you're going to do something else other than mapping after completing the current one (the script auto triggers and saves the log when entering a new map, or closing the script).

Left-clicking the movable tracker-button will open a window with available logs (sorted by days), displayed in a drop-down list -- right-clicking will pause the tracker altogether. Selecting a day will show all map runs from that day in a list: long-clicking these entries shows a tooltip with the tracked loot, holding right-click will delete that map run. There's also an export button in the top-right corner: left-clicking will export the data from the currently-selected day into a csv in the main script-folder, right-clicking will export all the data into a separate file.

I'm writing this kind of in a hurry because I'll be out for dinner shortly, so I may add or edit things I missed.

**Reply by GrimDawn-max** [reply]
I tested out a few things. First the good stuff

Map tracking and the timer work. The timer starts after entering the first portal. It will stop if I portal back out as expected, and resumes when I re-enter. That's great given you sometime do a trade while mapping. I didn't have side areas such as Vaal or other areas such as abyss or lab trials so couldn't test that.

I could move the star icon and the map timer bar around as mentioned. 
The export function works. I tested it after running multiple maps and all were included. The file export button worked (I neither left or right clicked. I just used the menu option to the right of the drop down). This showed all the map details with the loot column as "".

The stuff that doesn't appear to be working (for me at least)
The feature I could not get working was the loot inventory. I did the stash tab image check. I tried ctrl+left click with the inventory panel open and the map timer still running while in a map but this did not appear to do anything. I also tried ctrl+left click while in HO with the timer paused (not stopped as this stops tracking altogether). This didn't record anything either. There was no list generated to look at.

I might have misunderstood the instructions to get this working. If you can outline again or maybe put a short clip together to show me what to do I'll give it another shot tomorrow.

One last item - not related to the new features but I couldn't get the map properties to display. I selected three mods I always want to be alerted to and made them a 4. I zeroed out all the other mods. When I hovered over a map and pressed the scroll wheel (middle button) all I got were a list of the 4 native T16's the map could roll into with a horizon orb. There were no mods in the display. It still showed the example mod instead of the properties of the actual map I wanted to see. How do I get it so it shows the mods I want to be alerted to (they should show as a 4 in red I believe) ?

Overall so far so good !

**Reply by Lailloken** [reply]
https://www.youtube.com/watch?v=MwL6JfH9ejw

It should look like this. The stash needs to be open and the 'stash' plaque on top (the area that is screen-capped for the image-check) has to be visible. As for the map-info panel: do you have advanced item-info bound in-game to anything other than ALT? It's also the hotkey to show hidden loot on the ground. If you have bound it to something else, you need to configure the omni-key in the settings menu. There's a dedicated section for compatibility issues with custom in-game hotkeys.

EDIT:
Loot is being tracked if these conditions are met: loot tracker is enabled in the settings, a map is being tracked (the panel shows the name and timer), the stash is open, the item being ctrl-clicked is on the right side of the screen (it checks your cursor's position, so that clicks from the stash into inventory don't get tracked).

**Reply by GrimDawn-max** [reply]
Tested some more this morning. Got the loot tracker working after going back and re-reading the steps for screen capture on the wiki and doing them properly. The video was a handy reference to what it should look like. 

Entering a Lab trial paused the map timer as you expected. A purist might argue it should continue as that gives total time spent outside of the HO but but I prefer the way you have set it up as it is. 

Removing items worked perfectly, as did the export to csv file. The only quirky bit was that I had to make sure the map timer was stopped - not just paused - for the drop down menu for the export file to have all maps run displayed. If you don't do that, only the first map run stats appear. That might be obvious but I missed it initially.

I think it's all good to go !

On the other issue - map details display not working. I don't have anything else bound to ALT in-game. It's on the default. There are 2 issues. 

1. It doesn't seem to save my mod selection choices. It will only recognise the default options. They appear in the sub menu but never on the main one. How do I get my options to appear ?

2. I'm running the game in a Parallels based VM. Technically a Mac doesn't have an ALT key. The option key is mapped as ALT. This means I have to use ALT+LMB click to mimic the "normal" behaviour of ALT in-game on the windows system (the alt key works fine in the native MacOS POE client just not in a VM). That works fine but I can't remap the key in the UI in game within the VM to ALT+LMB click. It doesn't recognise it. I tried putting ALT in the Highlight section of the OMNI key but that didn't seem to do anything. Any ideas ? Here's a screenshot of what it looks like at the moment


<img width="1217" alt="MapInfoIssue" src="https://user-images.githubusercontent.com/64788480/197005145-f57f428c-0f0d-4991-acd0-a99895cb0db5.png">

**Reply by Lailloken** [reply]
> Entering a Lab trial paused the map timer as you expected. A purist might argue it should continue as that gives total time spent outside of the HO but but I prefer the way you have set it up as it is.

I think I'll add options and make it up to the user how side areas are handled. Technically, nothing is set up right now: it just checks if the current location is 'map' or not (well OK, I also added lakes).

> The only quirky bit was that I had to make sure the map timer was stopped - not just paused - for the drop down menu for the export file to have all maps run displayed. If you don't do that, only the first map run stats appear.

Is it really only the first, or the most recent one up to the current one? The tracker doesn't save anything until a map change is triggered, either manually by long-clicking the map name or automatically when the map changes or the script is closed. Until then, the logged data just sits in RAM, and the 'log viewer' doesn't have access.

Regarding the map-info:
I just checked something, and it's not an ALT issue as I first thought. So what exactly happens when you omni-click a map (just normal, not holding)? There should be a pop-up with either 'success', 'no mods' (if the mods on it are all set to 0), or 'not supported' (if the map is normal, unique, or un-ID).

**Reply by GrimDawn-max** [reply]
> I think I'll add options and make it up to the user how side areas are handled. Technically, nothing is set up right now: it just checks if the current location is 'map' or not (well OK, I also added lakes).

OK

>Is it really only the first, or the most recent one up to the current one? The tracker doesn't save anything until a map change is triggered, either manually by long-clicking the map name or automatically when the map changes or the script is closed. Until then, the logged data just sits in RAM, and the 'log viewer' doesn't have access.

I think it is the first map only but I only ran 2 maps to test it. Doesn't really matter. Now I know to stop the timer then all maps will be available to export is fine. Put the info on the wiki and everyone should be fine with it as it is.

>Regarding the map-info:
I just checked something, and it's not an ALT issue as I first thought. So what exactly happens when you omni-click a map (just normal, not holding)? There should be a pop-up with either 'success', 'no mods' (if the mods on it are all set to 0), or 'not supported' (if the map is normal, unique, or un-ID).

I tried it as you mentioned and it works as expected now. I think the issue was I was holding the omni-key too long. A simple one press of middle button was all I needed to do. The dangerous mods now appear as expected. 

I did run into another small issue. I edited the colour template file and saved the changes. This resulted in the mods showing up in a colour desired. However when I changed the colours again, the colours remained as the previous setting, not the new ones, even though I'd saved the changes again. Restarting the script didn't flush the old colour changes. Not a big deal and probably user error rather than an app issue. I might not have the game screen capture set properly amongst other things. If it really starts to bug me I'll look in to it further but for now it's all good. 

Thanks so much for these changes. They certainly help me a great deal and further the enjoyment of the game.

**Reply by Lailloken** [reply]
Since I found a method for color customization that I'm happy with in the item-info tooltip, I'll probably implement that for other features as well at some point.
If you find any other data you'd want the tracker to log as well (from the client.txt or otherwise), feel free to let me know. And thanks for your suggestion -- I found some useful UI elements and methods I can implement to improve user experience and visual feedback.

**Reply by GrimDawn-max** [reply]
>If you find any other data you'd want the tracker to log as well (from the client.txt or otherwise), feel free to let me know. And thanks for your suggestion -- I found some useful UI elements and methods I can implement to improve user experience and visual feedback.

The only other thing I can think of is in the Uber lab trials (Uber or the higher variants of it). Besides getting hopelessly lost in there even using POELab (which is half the point I guess) I always lose count of how many Dark shrines I've located. It would be good to know how many Dark shrines I've activated when I'm doing a lab run. 

I think you already pull some info for lab from POELab. You would need to check how many Dark shrines are in the daily (it's not always 4 on rare occasions) and then check the Client.txt for the comments that are stored there when you activate a shrine. I think there are only 4 comments Izaro makes. 

Another awesome feature in labs would be to keep track of which rooms you've already been in. That's not recorded anywhere in the Client.txt so not sure if that can be done. Then it would get marked on an overlay. Sometimes you have to back track in lab (especially the golden door) but most times it's a 1 and done. 

Can the map and loot trackers be used for lab runs ? It would be good to know how much time a lab run takes me as well as the loot I get, just like you've implemented for maps.

**Reply by Lailloken** [reply]
> Can the map and loot trackers be used for lab runs ? It would be good to know how much time a lab run takes me as well as the loot I get, just like you've implemented for maps.

I just checked the client.txt. Lab areas are tagged/named coherently, so I could add lab as an additional 'instance type' next to maps and lakes (and whatever else needs to be added, some boss instances are not tagged as maps).

> I think you already pull some info for lab from POELab.

My lab feature is pretty basic and old, which is why I want to update it at some point. It merely lets you import layout-images, overlay them on the client, and set a marker on it for orientation. I don't know how advanced I want it to be (it needs some UI and UX improvements at the very least) because I don't particularly like running lab. It's just there to get rid of alt-tabbing. The most straight-forward implementation of navigation/tracking would be to expand the marker system so that one may place markers on every room one wants to visit -- plan a path, so to speak. Once rooms are cleared, one could manually mark them as completed with color highlighting or an icon (this could be done during transitions/loading-screens).

**Reply by GrimDawn-max** [reply]
>> I just checked the client.txt. Lab areas are tagged/named coherently, so I could add lab as an additional 'instance type' next to maps and lakes (and whatever else needs to be added, some boss instances are not tagged as maps).

Sounds good. The loot tracking isn't important but a timer for the lab, that can be stored in a csv for each run to look at later  would be good.

>> My lab feature is pretty basic and old, which is why I want to update it at some point. It merely lets you import layout-images, overlay them on the client, and set a marker on it for orientation. I don't know how advanced I want it to be (it needs some UI and UX improvements at the very least) because I don't particularly like running lab. It's just there to get rid of alt-tabbing. The most straight-forward implementation of navigation/tracking would be to expand the marker system so that one may place markers on every room one wants to visit -- plan a path, so to speak. Once rooms are cleared, one could manually mark them as completed with colour highlighting or an icon (this could be done during transitions/loading-screens).

I suck at running labs and only do so to get currency. I get lost frequently and it ends up taking me about 30 minutes to complete one, which is about 15 times slower than lab runner. If something helped me get it down to 10 minutes I'd be happy - I don't need to go at those insane speeds. There is a program called labbie that is supposed to help with tracking but I've not tried it. Ideally what I'd like it a min-map that show you which way to go to get to the entrances to room. There's a file that goes with POELab that is supposed to help with that but all it really does it tell you to go left or right from your entrance point. If you implement the time tracker for labs I'll be happy. The other stuff would be icing on the cake but there's no hurry for it.

**Reply by Lailloken** [reply]
Finally found some time to continue working on this, and I found a significant bug that crept in last minute before I uploaded the test version:

> You can long-click the map name to manually trigger a map as completed -- this is only needed if you know you're going to do something else other than mapping after completing the current one (the script auto triggers and saves the log when entering a new map, or closing the script).

The auto-trigger when entering a new map wasn't working for some reason. The normal behavior would be:
- start the client, start script, tracker shows 'not tracking'
- enter map, tracker starts the timer, finish map, stash loot
- open new map, enter, log is saved, tracker resets with the new map and the timer at 00:00

The bug caused the tracker to go back to 'not tracking' and 00:00 when entering a new map, making it necessary to re-enter the map in order to start tracking (or manually setting the previous map as completed before entering a new one). Just a heads-up for when the stable version is released.

**Reply by GrimDawn-max** [reply]
OK. Haven't come across it myself but my testing was fairly limited.
